flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/slot/TaskSlotTableImplTest.java

```diff
@@ -137,6 +137,68 @@ public void testRedundantSlotAllocation() throws Exception {
         }
     }
 
+    /**
+     * Tests that inconsistent dynamic slot allocation with the same AllocationID to a different
+     * slot is rejected.
+     */
+    @Test
+    public void testInconsistentDynamicSlotAllocation() throws Exception {
+        try (final TaskSlotTable<TaskSlotPayload> taskSlotTable = createTaskSlotTableAndStart(1)) {
+            final JobID jobId1 = new JobID();
+            final JobID jobId2 = new JobID();
+            final AllocationID allocationId = new AllocationID();
+
+            assertThat(
+                    taskSlotTable.allocateSlot(-1, jobId1, allocationId, SLOT_TIMEOUT), is(true));
+            assertThat(
+                    taskSlotTable.allocateSlot(-1, jobId2, allocationId, SLOT_TIMEOUT), is(false));
+
+            assertThat(taskSlotTable.isAllocated(-1, jobId1, allocationId), is(true));
+
+            Iterator<TaskSlot<TaskSlotPayload>> allocatedSlots =
+                    taskSlotTable.getAllocatedSlots(jobId1);
+            assertThat(allocatedSlots.next().getAllocationId(), is(allocationId));
+            assertThat(allocatedSlots.hasNext(), is(false));
+        }
+    }
+
+    @Test
+    public void testDuplicateStaticSlotAllocation() throws Exception {
+        try (final TaskSlotTable<TaskSlotPayload> taskSlotTable = createTaskSlotTableAndStart(2)) {
+            final JobID jobId = new JobID();
+            final AllocationID allocationId = new AllocationID();
+
+            assertThat(taskSlotTable.allocateSlot(0, jobId, allocationId, SLOT_TIMEOUT), is(true));
+            assertThat(taskSlotTable.allocateSlot(0, jobId, allocationId, SLOT_TIMEOUT), is(true));
+
+            assertThat(taskSlotTable.isAllocated(0, jobId, allocationId), is(true));
+            assertThat(taskSlotTable.isSlotFree(1), is(true));
+
+            Iterator<TaskSlot<TaskSlotPayload>> allocatedSlots =
+                    taskSlotTable.getAllocatedSlots(jobId);
+            assertThat(allocatedSlots.next().getIndex(), is(0));
+            assertThat(allocatedSlots.hasNext(), is(false));
+        }
+    }
+
+    @Test
+    public void testDuplicateDynamicSlotAllocation() throws Exception {
```

Comment:
I think we should also verify that the second call of `allocateSlot` does not replace the slot allocated by the first call.